<a href="https://colab.research.google.com/github/pedrohsrmorais/Machine-Learning-para-Churn-de-clientes/blob/master/Desafio_6_DNC_Pedro_Henrique.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.preprocessing import LabelEncoder

dataframe = pd.read_csv('streaming_data.csv', sep=',',encoding='latin-1')
dataframe.head(6)


Age  Gender  Time_on_platform  Devices_connected Subscription_type  \
0  49.0    Male            4700.0                3.0             Basic   
1  69.0  Female            6107.0                2.0          Standard   
2  44.0    Male            6374.0                3.0           Premium   
3  26.0    Male               NaN                5.0             Basic   
4  18.0     NaN            1948.0                3.0             Basic   
5  54.0    Male            3024.0                2.0             Basic   

   Num_streaming_services  Num_active_profiles  Avg_rating  Churned  \
0                     NaN                    3         NaN      0.0   
1                     NaN                    1        2.14      0.0   
2                     1.0                    1        1.91      1.0   
3                     NaN                    2         NaN      NaN   
4                     NaN                    3        3.88      0.0   
5                     NaN                    1         NaN      0.0   

                                User_id  
0  a271aa56-bcfc-4f0a-91f7-b773de6b86a4  
1  fe14b048-770b-4619-8ec6-485b0969ae31  
2  0283dffd-6684-4a05-9c6f-c02098a6849d  
3  c316f2f5-3403-4a0a-82c2-c98e4b3379d2  
4  14879bcb-1977-4ad8-a7f9-6aa9bf7b872f  
5  a1df3a13-9255-4d00-8a9d-20565fefaab9

In [ ]:
# Verificando e removendo dados nulos
print("número de Dados nulos: ")
print(dataframe.isna().sum())

# Substituir valores NaN por 0 em colunas específicas
dataframe['Time_on_platform'].fillna(0, inplace=True)
dataframe['Num_streaming_services'].fillna(0, inplace=True)
dataframe['Churned'].fillna(0, inplace=True)
dataframe['Avg_rating'].fillna(0, inplace=True)
dataframe['Devices_connected'].fillna(0, inplace=True)


# Dropando dados nulos em gender, subscription_type e age
dataframe.dropna(subset=['Gender', 'Subscription_type', 'Age'], inplace=True)

número de Dados nulos: 
Age                       11277
Gender                    14516
Time_on_platform          21420
Devices_connected         25188
Subscription_type         25221
Num_streaming_services    68312
Num_active_profiles           0
Avg_rating                17270
Churned                   16736
User_id                       0
dtype: int64


In [ ]:
print(dataframe.describe())

dataframe.info()

# Transformações para 'Age'
dataframe['Age'] = pd.to_numeric(dataframe['Age'], errors='coerce', downcast='integer')

# Transformações para 'Gender'
dataframe['Gender'] = LabelEncoder().fit_transform(dataframe['Gender'])

# Transformações para 'Devices_connected'
dataframe['Devices_connected'] = pd.to_numeric(dataframe['Devices_connected'], errors='coerce', downcast='integer')

# Transformações para 'Subscription_type'
dataframe['Subscription_type'] = LabelEncoder().fit_transform(dataframe['Subscription_type'])

# Transformações para 'Num_streaming_services'
dataframe['Num_streaming_services'] = pd.to_numeric(dataframe['Num_streaming_services'], errors='coerce', downcast='integer')

# Transformações para 'Num_active_profiles'
dataframe['Num_active_profiles'] = pd.to_numeric(dataframe['Num_active_profiles'], errors='coerce', downcast='integer')

# Transformações para 'User_id'
dataframe['User_id'] = dataframe['User_id'].astype(str)

dataframe.head(10)

                Age  Time_on_platform  Devices_connected  \
count  36515.000000      36515.000000       36515.000000   
mean      43.473422       3170.231576           2.039628   
std       15.092181       2909.245246           1.820622   
min       18.000000          0.000000           0.000000   
25%       30.000000          0.000000           0.000000   
50%       43.000000       2708.000000           2.000000   
75%       57.000000       5741.000000           4.000000   
max       69.000000       8759.000000           5.000000   

       Num_streaming_services  Num_active_profiles    Avg_rating       Churned  
count            36515.000000         36515.000000  36515.000000  36515.000000  
mean                 0.309434             2.993455      2.337959      0.193236  
std                  0.909600             1.412968      1.609043      0.394842  
min                  0.000000             1.000000      0.000000      0.000000  
25%                  0.000000             2.000000    

Age  Gender  Time_on_platform  Devices_connected  Subscription_type  \
0    49       1            4700.0                  3                  0   
1    69       0            6107.0                  2                  2   
2    44       1            6374.0                  3                  1   
3    26       1               0.0                  5                  0   
5    54       1            3024.0                  2                  0   
9    21       0             112.0                  2                  0   
11   61       1            5389.0                  3                  2   
14   44       1            4198.0                  1                  0   
15   30       1            2306.0                  3                  1   
20   20       1            2542.0                  5                  0   

    Num_streaming_services  Num_active_profiles  Avg_rating  Churned  \
0                        0                    3        0.00      0.0   
1                        0                    1        2.14      0.0   
2                        1                    1        1.91      1.0   
3                        0                    2        0.00      0.0   
5                        0                    1        0.00      0.0   
9                        0                    3        2.57      1.0   
11                       0                    4        3.92      0.0   
14                       0                    3        2.59      0.0   
15                       0                    5        3.95      0.0   
20                       0                    1        0.00      1.0   

                                 User_id  
0   a271aa56-bcfc-4f0a-91f7-b773de6b86a4  
1   fe14b048-770b-4619-8ec6-485b0969ae31  
2   0283dffd-6684-4a05-9c6f-c02098a6849d  
3   c316f2f5-3403-4a0a-82c2-c98e4b3379d2  
5   a1df3a13-9255-4d00-8a9d-20565fefaab9  
9   c6f0b9be-dd53-42a3-8efd-1ad7dcdda167  
11  8f62edd7-4fe0-45c2-98b9-2951e4bef897  
14  5417ca3a-b300-4429-b227-3bc75fbb2c1d  
15  cc4de3a2-a732-42f1-b55d-6e33886a3b7d  
20  b4da071a-c8b4-44dd-84fb-7de32abf766f

In [ ]:
print("A média dos ratings é: ", dataframe['Avg_rating'].mean())

# Tratamento dos ratings (maior que 5, menor que 0):
valores_invalidos = dataframe[(dataframe['Avg_rating'] > 5) | (dataframe['Avg_rating'] < 0)]['Avg_rating']

# Exibir os resultados
print("Valores inválidos (maiores que 5 ou menores que 0):", valores_invalidos)

A média dos ratings é:  2.337959468711489
Valores inválidos (maiores que 5 ou menores que 0): Series([], Name: Avg_rating, dtype: float64)


In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler

# Suponha que dataframe seja o seu DataFrame com os dados
# Definição das variáveis X e y
X = dataframe.drop(['Churned', 'User_id'], axis=1)
y = dataframe['Churned']

# Dividir os dados em conjuntos de treinamento e teste
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Definir o pipeline com pré-processamento e modelo
pipeline = Pipeline([
    ('scaler', StandardScaler()),  # Normalização dos dados
    ('clf', LogisticRegression())  # Modelo de regressão logística
])

# Definir os parâmetros para ajustar
parameters = {
    'clf__C': [0.001, 0.01, 0.1, 1, 10, 100],  # Parâmetro de regularização
    'clf__penalty': ['l1', 'l2']  # Tipo de regularização
}

# Realizar a pesquisa em grade
grid_search = GridSearchCV(pipeline, parameters, cv=5, n_jobs=-1)

# Treinar o modelo com a pesquisa em grade
grid_search.fit(X_train, y_train)

# Exibir os melhores parâmetros encontrados
print("Melhores parâmetros encontrados:")
print(grid_search.best_params_)

# Fazer previsões com o modelo ajustado
y_pred = grid_search.predict(X_test)

# Avaliar o modelo ajustado
accuracy = accuracy_score(y_test, y_pred)
print("Acurácia:", accuracy)

# Outras métricas
print(classification_report(y_test, y_pred))
print("Matriz de Confusão:")
print(confusion_matrix(y_test, y_pred))


Melhores parâmetros encontrados:
{'clf__C': 0.001, 'clf__penalty': 'l2'}
Acurácia: 0.8078871696563056
              precision    recall  f1-score   support

         0.0       0.81      1.00      0.89      5900
         1.0       0.00      0.00      0.00      1403

    accuracy                           0.81      7303
   macro avg       0.40      0.50      0.45      7303
weighted avg       0.65      0.81      0.72      7303

Matriz de Confusão:
[[5900    0]
 [1403    0]]


/usr/local/lib/python3.10/dist-packages/sklearn/model_selection/_validation.py:378: FitFailedWarning: 
30 fits failed out of a total of 60.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
30 fits failed with the following error:
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/sklearn/model_selection/_validation.py", line 686, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/usr/local/lib/python3.10/dist-packages/sklearn/pipeline.py", line 405, in fit
    self._final_estimator.fit(Xt, y, **fit_params_last_step)
  File "/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py", line 1162, in fit
    solver = _check_solver(self.solver, self.penalty, self.dual)
 

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.model_selection import GridSearchCV
from sklearn.preprocessing import MinMaxScaler


# Construir o modelo de regressão logística
model = RandomForestClassifier()

# Separando os dados em variáveis independentes (X) e variável dependente (y)
X = dataframe.drop(['Churned', 'User_id'], axis=1)
y = dataframe['Churned']

# Criar uma instância do MinMaxScaler
scaler = MinMaxScaler()

# Dividir os dados em conjuntos de treinamento e teste
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Ajustar o scaler aos dados de treinamento (calcula o mínimo e o máximo para cada recurso)
scaler.fit(X_train)

# Transformar os dados de treinamento e teste usando o scaler ajustado
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Treinar o modelo com os dados normalizados
model.fit(X_train_scaled, y_train)

# Fazer previsões com os dados de teste normalizados
y_pred = model.predict(X_test_scaled)

# Avaliar o modelo
accuracy = accuracy_score(y_test, y_pred)
print("Acurácia:", accuracy)

# Outras métricas
print(classification_report(y_test, y_pred))
print("Matriz de Confusão:")
print(confusion_matrix(y_test, y_pred))


Acurácia: 0.8504724085992058
              precision    recall  f1-score   support

         0.0       0.85      0.99      0.91      5900
         1.0       0.87      0.26      0.40      1403

    accuracy                           0.85      7303
   macro avg       0.86      0.63      0.66      7303
weighted avg       0.85      0.85      0.82      7303

Matriz de Confusão:
[[5847   53]
 [1039  364]]


A Acurácia do modelo Random Forest foi melhor durante os testes, tendo cerca de 0.85 e portanto seria o modelo a ser escolhido para prever a chance de retenção de determinado cliente.